src/copyable_task.jl

```diff
@@ -0,0 +1,1139 @@
+ways. For example, in the context of Sequential Monte Carlo, you might want the only
+difference between two copies to be their random number generator.
+
+A generic mechanism is available to achieve this. [`Libtask.get_taped_globals`](@ref) and
+[`Libtask.set_taped_globals!`](@ref) let you set and retrieve a variable which is specific
+to a given [`Libtask.TapedTask`](@ref). The former can be called inside a function:
+```jldoctest sv
+julia> function f()
+           produce(get_taped_globals(Int))
+           produce(get_taped_globals(Int))
+           return nothing
+       end
+f (generic function with 1 method)
+```
+
+The first argument to [`Libtask.TapedTask`](@ref) is the value that
+[`Libtask.get_taped_globals`](@ref) will return:
+```jldoctest sv
+julia> t = TapedTask(1, f);
+
+julia> consume(t)
+1
+```
+
+The value that it returns can be changed between [`Libtask.consume`](@ref) calls:
+```jldoctest sv
+julia> set_taped_globals!(t, 2)
+
+julia> consume(t)
+2
+```
+
+`Int`s have been used here, but it is permissible to set the value returned by
+[`Libtask.get_taped_globals`](@ref) to anything you like.
+
+# Implementation Notes
+
+Under the hood, we implement a `TapedTask` by obtaining the `IRCode` associated to the
+original function, transforming it so that it implements the semantics required by the
+`produce` / `consume` interface, and placing it inside a `MistyClosure` to make it possible
+to execute.
+
+There are two main considerations when transforming the `IRCode`. The first is to ensure
+that the "state" of a `TapedTask` can be copied, so that a `TapedTask` can be copied, and
+resumed later. The complete state of a `TapedTask` is given by its arguments, and the value
+associated to each ssa (these are initially undefined).
+To make it possible to copy the state of the ssa values, we place a `Base.RefValue{T}`s into
+the captures of the `MistyClosure` which implements the `TapedTask`, one for each ssa in the
+IR (`T` is the type inferred for that ssa). A call is replaced by reading in values of ssas
+from these refs, applying the original operation, and writing the result to the ref
+associated to the instruction. For example, if the original snippet of `IRCode` is something
+like
+```julia
+%5 = f(%3, _1)
+```
+the transformed IR would be something like
+```julia
+%5 = ref_for_%3[]
+%6 = f(%5, _1)
+     ref_for_%5[] = %6
+```
+Setting things up in this manner ensures that an independent copy is made by simply copying
+all of the refs. A `deepcopy` is required for correctness as, while the refs to not alias
+one another (by construction), their contents might. For example, two refs may contain the
+same `Array`, and in general the behaviour of a function depends on this relationship.
+
+The second component of the transformation is implementing the `produce` mechanism, and the
+ability to resume computation from where we produced. Roughly speaking, the `IRCode` must be
```

Comment:
A hint on how resumption is implemented helps demystify what happens under the hood. I'm not sure what I wrote is accurate, so feel free to improve it. 

```suggestion
ability to resume computation from where we produced (via [`goto` ](https://docs.julialang.org/en/v1/base/base/#Base.@goto)statements). Roughly speaking, the `IRCode` must be
```

Reply:
Ah, right, yes. It's similar, but not quite the same mechanism as the `@goto` statement. I've tweaked the docstring accordingly.